sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -568,6 +568,29 @@ public Duration getAllowedTimestampSkew() {
   public @interface ProcessElement {}
 
   /**
+   * <b><i>Experimental - no backwards compatibility guarantees. The exact name or usage of this
+   * feature may change.</i></b>
+   *
+   * <p>Annotation that may be added to a {@link ProcessElement} or {@link OnTimer} method to
+   * indicate that the observable contents of the input {@link PCollection} or mutable state must be
+   * stable upon retries.
+   *
+   * <p>This is important for exactly-once semantics when writing to a storage medium outside of
+   * your pipeline. The general pattern is to establish an idempotent protocol for writing the data,
+   * and then require stable input. Combined, these allow the write to be freely retried until
+   * success.
+   *
+   * <p>An example of an unstable input would be anything computed using nondeterministic logic. In
+   * Beam, any user-defined function is permitted to be nondeterministic, and any {@link
+   * PCollection} is permitted to be recomputed in any manner.
+   */
```

Comment:
What about linking to an example IO sink in this documentation to demonstrate its intended use?

Reply:
After this is in, I will file a bunch of TODOs to add it where it is needed in other IOs.